lib/puppet/etc.rb

```diff
@@ -0,0 +1,118 @@
+    def getgrent
+      if group_entry = Etc.getgrent
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # closes handle to /etc/group file
+    def endgrent
+      Etc.endgrent
+    end
+
+    # effectively equivalent to IO#rewind of /etc/group
+    def setgrent
+      Etc.setgrent
+    end
+
+    # Etc::getpwent returns an Etc::Passwd struct object
+    # On first call opens /etc/passwd and returns parse of first entry. Each subsquent call
+    # returns new struct for the next entry or nil if EOF. Call ::endgrent to close file.
+    def getpwent
+      if user_entry = Etc.getpwent
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    # closes handle to /etc/passwd file
+    def endpwent
+      Etc.endpwent
+    end
+
+    #effectively equivalent to IO#rewind of /etc/passwd
+    def setpwent
+      Etc.setpwent
+    end
+
+    # Etc::getpwnam searches /etc/passwd file for an entry corresponding to
+    # username.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwnam(username)
+      if user_entry = Etc.getpwnam(username)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    # Etc::getgrnam searches /etc/group file for an entry corresponding to groupname.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrnam(groupname)
+      if group_entry = Etc.getgrnam(groupname)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getgrid searches /etc/group file for an entry corresponding to id.
+    # returns an Etc::Group struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getgrgid(id)
+      if group_entry = Etc.getgrgid(id)
+        convert_field_values_to_utf8!(group_entry)
+      end
+      group_entry
+    end
+
+    # Etc::getpwuid searches /etc/passwd file for an entry corresponding to id.
+    # returns an Etc::Passwd struct corresponding to the entry or raises
+    # ArgumentError if none
+    def getpwuid(id)
+      if user_entry = Etc.getpwuid(id)
+        convert_field_values_to_utf8!(user_entry)
+      end
+      user_entry
+    end
+
+    private
+    # utility method for converting the values of a struct returned by the Etc
+    # module to UTF-8.
+    #
+    # Warning! This is a destructive method - the struct passed is modified!
+    #
+    # Structs returned by the ruby Etc module contain members with
+    # fields of type String, Integer, or Array of Strings, so we handle these
+    # types. Otherwise ignore fields.
+    def convert_field_values_to_utf8!(struct)
+      struct.each_with_index do |value, index|
+        if value.is_a?(String)
+          struct[index] = Puppet::Util::CharacterEncoding.convert_to_utf_8(value)
```

Comment:
Thanks for pointing that out @branan. I'm not sure if this is the right thing to do.. I'm very slightly inclined to say it would be more surprising if puppet managed some properties of a user and failed to manage others (assuming these are supported properties under management) and as a user I would rather have puppet fail to manage the user than fail to manage some parts of it. The case I'm thinking of is something like group membership, where we might create a new user and add it some unprivileged UTF-8 named group, or perhaps want to remove an existing user from a privileged UTF-8 named group. I think I'd rather have the user not created than have it created and put in incorrect groups. Wouldn't mind a second opinion from @puppet-maintainers..